CMakeLists.txt

```diff
@@ -52,39 +60,105 @@ else(NOT DISABLE_LIBSODIUM_RNG_SEED_FUNCTION)
         SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DDISABLE_LIBSODIUM_RNG_SEED_FUNCTION")
 endif(NOT DISABLE_LIBSODIUM_RNG_SEED_FUNCTION)
 
-set(ECDAA_SRC_DIR src)
-set(ECDAA_INTERNAL_SRC_DIR ${ECDAA_SRC_DIR}/internal/)
-set(ECDAA_AMCL_EXTENSIONS_SRC_DIR ${ECDAA_SRC_DIR}/amcl-extensions/)
-set(ECDAA_SRCS
-        ${ECDAA_SRC_DIR}/credential_BN254.c
-        ${ECDAA_SRC_DIR}/group_public_key_BN254.c
-        ${ECDAA_SRC_DIR}/issuer_keypair_BN254.c
-        ${ECDAA_SRC_DIR}/member_keypair_BN254.c
-        ${ECDAA_SRC_DIR}/prng.c
-        ${ECDAA_SRC_DIR}/signature_BN254.c
-        ${ECDAA_INTERNAL_SRC_DIR}/explicit_bzero.c
-        ${ECDAA_INTERNAL_SRC_DIR}/schnorr.c
-        ${ECDAA_AMCL_EXTENSIONS_SRC_DIR}/big_256_56.c
-        ${ECDAA_AMCL_EXTENSIONS_SRC_DIR}/ecp_BN254.c
-        ${ECDAA_AMCL_EXTENSIONS_SRC_DIR}/ecp2_BN254.c
-        ${ECDAA_AMCL_EXTENSIONS_SRC_DIR}/pairing_BN254.c
+set(GENERATED_TOPLEVEL_INCLUDE_DIR "${CMAKE_BINARY_DIR}/include")
+
+set(BIG_PATTERN "XXX")
+set(FP_PATTERN "YYY")
+set(CURVE_PATTERN "ZZZ")
+
+file(GLOB PUBLIC_HEADERS "include/ecdaa/*.h")
+file(GLOB_RECURSE IMPLEMENTATION_H_FILES "src/*.h")
+file(GLOB_RECURSE IMPLEMENTATION_C_FILES "src/*.c")
+file(GLOB_RECURSE TEST_C_FILES "test/*.c")
+file(GLOB_RECURSE TEST_H_FILES "test/*.h")
+set(FILES_TO_PROCESS ${PUBLIC_HEADERS} ${IMPLEMENTATION_H_FILES} ${IMPLEMENTATION_C_FILES} ${TEST_C_FILES} ${TEST_H_FILES})
+
+foreach(curve ${ECDAA_CURVES})
```

Comment:
Doing the template expansion in CMake like this means that template is only expanded one during setup (`cmake ..`) and not on each build (`make`).  A modified template file is not re-expanded during a normal build.

It would be better to define these as build targets.  The `Makefile` generated by `cmake ..` would contain something like

```make
src/credentials_BN254.c : src/credentials_ZZZ.c
       my_expand_command $< BN254

src/group_public_key_BLS383.c: src/group_public_key_ZZZ.c
        my_exapnd_command $< BLS383

...
```

Reply:
The `my_expand_command` would need to be something other than CMake.  Since AMCL is already using Python3, that might be a good choice.

Reply:
Rushed this comment, initially.

The pragmatic thing here might be to keep this commit as is, and open an issue for reworking the template/build system interaction.  We could burn too much time trying to address this now.

Reply:
This is a good point.

I actually wanted to have the processing done at configure time, so the files would get setup without having to run the build. But, yea, that's a stupid setup for development. I didn't realize this, because I'm in the habit of re-running the configure step for each build.

So, yes, I agree.

I spent all morning working with our code-coverage tools to make them make sense with all the generated code, so my day is shot. I'm going to see if I can't just make this conversion to python this afternoon and be done with it.